cadence/contracts/DapperStorageRent.cdc

```diff
@@ -0,0 +1,247 @@
+import FungibleToken from "./FungibleToken.cdc"
+import FlowToken from "./FlowToken.cdc"
+import PrivateReceiverForwarder from "./PrivateReceiverForwarder.cdc"
+
+/// DapperStorageRent
+/// Provide a means for accounts storage TopUps. To be used during transaction execution.
+pub contract DapperStorageRent {
+
+  pub let DapperStorageRentAdminStoragePath: StoragePath
+
+  /// Threshold of storage required to trigger a refill
+  access(contract) var StorageRentRefillThreshold: UInt64
+  /// List of all refilledAccounts
+  access(contract) var RefilledAccounts: [Address]
+  /// Detailed account information of refilled accounts
+  access(contract) var RefilledAccountInfos: {Address: RefilledAccountInfo}
+  /// List of all blockedAccounts
+  access(contract) var BlockedAccounts: [Address]
+  /// Blocks required between refill attempts
+  access(contract) var RefillRequiredBlocks: UInt64
+
+  /// Event emitted when an Admin blocks an address
+  pub event BlockedAddress(_ address: [Address])
+  /// Event emitted when a Refill is successful
+  pub event Refuelled(_ address: Address)
+  /// Event emitted when a Refill is not successful
+  pub event RefilledFailed(address: Address, reason: String)
+
+  /// getStorageRentRefillThreshold
+  /// Get the current StorageRentRefillThreshold
+  ///
+  /// @return UInt64 value of the current StorageRentRefillThreshold value
+  pub fun getStorageRentRefillThreshold(): UInt64 {
+    return self.StorageRentRefillThreshold
+  }
+
+  /// getRefilledAccounts
+  /// Get the current StorageRentRefillThreshold
+  ///
+  /// @return List of refilled Accounts
+  pub fun getRefilledAccounts(): [Address] {
+    return self.RefilledAccounts
+  }
+
+  /// getBlockedAccounts
+  /// Get the current StorageRentRefillThreshold
+  ///
+  /// @return List of blocked accounts
+  pub fun getBlockedAccounts() : [Address] {
+    return self.BlockedAccounts
+  }
+
+  /// getRefilledAccountInfos
+  /// Get the current StorageRentRefillThreshold
+  ///
+  /// @return Address: RefilledAccountInfo mapping
+  pub fun getRefilledAccountInfos(): {Address: RefilledAccountInfo} {
+    return self.RefilledAccountInfos
+  }
+
+  /// getRefillRequiredBlocks
+  /// Get the current StorageRentRefillThreshold
+  ///
+  /// @return UInt64 value of the current RefillRequiredBlocks value
+  pub fun getRefillRequiredBlocks(): UInt64 {
+    return self.RefillRequiredBlocks
+  }
+
+  /// tryRefill
+  /// Attempt to refill an accounts storage capacity if it has dipped below threshold and passes other checks.
+  ///
+  /// @param address: Address to attempt a storage refill on
+  pub fun tryRefill(_ address: Address) {
```

Comment:
Keep in mind that it can be called by anyone onchain which means anyone can suck all the funds out of it.

My recommendation is that it should only be allowed to call by the dapper key / more like a multi-signature where the payer would be the token receiver.